Refactor kubernetes autodiscover to avoid skipping short-living pods

jsoriano opened the pull request:
## What does this PR do?

Refactor logic in kubernetes autodiscover that decides when to generate events to try to address #22718.

Kubernetes autodiscover can generate events without network information now (without `host` or `port`/`ports`). This allows to generate events for pods that haven't started yet, or have succeeded/failed before generating a running event. These events still include the container id, so they can be used to collect logs. Still, no start event is generated if no pod ip and no container ids are available.

Some helpers have been added to obtain relevant information from pods and their containers.

Some additional small refactors are done to improve readability.

## Why is it important?

Current logic is checking similar things at the pod and container levels, try to simplify this logic focusing in containers only.

No matter what is the state of the pod, if there is a container running or trying to run, even if it is unhealthy, some configuration should be generated, so logs can be collected.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] ~~I have made corresponding changes to the documentation~~
- [x] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works. But more tests would be needed.
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [x] Fix update so it definitely stops configurations, respecting the cleanup timeout.
- [x] Double-check that nothing breaks with events without `data.host`.
  This is logged at the debug level:
  ```
  2021-04-08T12:53:34.453+0200	DEBUG	[autodiscover]	template/config.go:156	Configuration template cannot be resolved: field 'data.host' not available in event or environment accessing 'hosts' (source:'/home/jaime/tmp/metricbeat-autodiscover-k8s.yml')
  ```
- [x] Actually check that the refactor solves the issues and doesn't break anything else.
- [x] Try to add more testing. A couple of cases added to current tests.
- [x] Check that nothing breaks with hints including `data.host` or `data.ports`.
  - [x] `data.ports.*` doesn't seem to be working, check if this is a regression.  Named ports were not added to container events. They are now.
  - [x] Multiple configurations generated for pods with multiple containers, expected? compare with master. False alarm.
- [x] Check autodiscover with heartbeat.

## Author's notes for the future

- When using autodiscover templates, if the condition matches pod conditions, container events will match too because they also include pod metadata. In pods with multiple containers this may lead to events with metadata of the incorrect container. This could be possibly avoided by matching per pod metadata only in the pod events. This is also happening in master, not solving in this PR. 
- `data.ports` is not included in container-level events, so it doesn't work when conditions for specific containers are used.

## How to test this PR locally

* Check that everything keeps running with some autodiscover happy cases.
* Check that logs are collected from a pod in CrashLoopBackOff (`kubectl run --image=redis redis -- echo foo`).

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #22718.

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->
Collect logs from containers in short-living or failing pods.

elasticmachine commented:
## :broken_heart: Tests Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-24742/detail/PR-24742/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-24742/detail/PR-24742/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-24742/detail/PR-24742/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-24742/detail/PR-24742/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_24742.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #24742 opened`



* Start Time: 2021-03-24T18:24:40.176+0000


* Duration: 109 min 55 sec


* Commit: bc37e731691c99d1a02b9e1e1ec808a18f95ba4e



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 3  |
| Passed       | 44774  |
| Skipped      | 4737 |
| Total        | 49514   |


#### Trends :test_tube:


![Image of Build Times](https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-24742/buildTimeGraph/png)


![Image of Tests](https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-24742/1/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![3](https://img.shields.io/badge/3%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-24742/detail/PR-24742/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `Build&Test / filebeat-windows-2016-windows-2016 / TestDefaultHarvesterGroup/assert_a_harvester_can_be_stopped_and_removed_from_bookkeeper – github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestDefaultHarvesterGroup/assert_a_harvester_can_be_stopped_and_removed_from_bookkeeper
    harvester_test.go:108: 
        	Error Trace:	harvester_test.go:108
        	            				harvester_test.go:151
        	Error:      	Should be false
        	Test:       	TestDefaultHarvesterGroup/assert_a_harvester_can_be_stopped_and_removed_from_bookkeeper
    --- FAIL: TestDefaultHarvesterGroup/assert_a_harvester_can_be_stopped_and_removed_from_bookkeeper (0.00s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / filebeat-windows-2016-windows-2016 / TestDefaultHarvesterGroup – github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestDefaultHarvesterGroup
--- FAIL: TestDefaultHarvesterGroup (0.03s)
 
 ```
</p></details>
</ul>
    
##### `Build&Test / auditbeat-windows-8-windows-8 / TestNonRecursive – github.com/elastic/beats/v7/auditbeat/module/file_integrity/monitor`


<ul>
<details><summary>Expand to view the error details</summary><p>

```
 Failed 
 ```
</p></details>
<details><summary>Expand to view the stacktrace</summary><p>

```
 === RUN   TestNonRecursive
    monitor_test.go:73: 
        	Error Trace:	monitor_test.go:73
        	Error:      	An error is expected but got nil.
        	Test:       	TestNonRecursive
    monitor_test.go:74: 
        	Error Trace:	monitor_test.go:74
        	Error:      	Not equal: 
        	            	expected: *errors.errorString(&errors.errorString{s:"read timeout"})
        	            	actual  : <nil>(<nil>)
        	Test:       	TestNonRecursive
--- FAIL: TestNonRecursive (3.02s)
 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![9](https://img.shields.io/badge/9%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-24742/detail/PR-24742/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `auditbeat-windows-8-windows-8 - mage build unitTest`
<ul>
<li>Took 5 min 31 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/12687/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `gsutil -m -q cp -a public-read test-build-artifacts-auditbeat-windows-8-windows-8-tgz gs://beats-ci-`
<ul>
<li>Took 0 min 23 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/19326/log/?start=0">here</a></li>
<li>Description: <code>@echo off                    gsutil -m -q cp -a public-read test-build-artifacts-auditbeat-windows-8-windows-8-tgz gs://beats-ci-temp/Beats/beats/PR-24742-1</code></l1>
</ul>
  
  
  
  
##### `filebeat-windows-2016-windows-2016 - mage build unitTest`
<ul>
<li>Took 3 min 18 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/9772/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `gsutil -m -q cp -a public-read test-build-artifacts-filebeat-windows-2016-windows-2016-tgz gs://beat`
<ul>
<li>Took 1 min 17 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/15528/log/?start=0">here</a></li>
<li>Description: <code>@echo off                    gsutil -m -q cp -a public-read test-build-artifacts-filebeat-windows-2016-windows-2016-tgz gs://beats-ci-temp/Beats/beats/PR-24742-1</code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 5 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/2822/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 1 min 13 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/6657/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 1 min 19 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/1814/log/?start=0">here</a></li>
<li>Description: <code>[2021-03-24T18:38:04.259Z] [Google Cloud Storage Plugin] Found 1 files to download from pattern: gs:</code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 12 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/4514/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-24742/runs/1/steps/13627/log/?start=0">here</a></li>
<li>Description: <code>Error 'hudson.AbortException: script returned exit code 1'</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-03-24T20:12:46.854Z] 11.09s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_003_panw
[2021-03-24T20:12:46.854Z] 11.02s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_214_ibmmq
[2021-03-24T20:12:46.854Z] 10.65s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_258_juniper
[2021-03-24T20:12:46.854Z] 10.61s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_028_oracle
[2021-03-24T20:12:46.854Z] 10.60s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_018_oracle
[2021-03-24T20:12:46.854Z] 10.58s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_006_squid
[2021-03-24T20:12:46.854Z] 10.50s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_147_fortinet
[2021-03-24T20:12:46.854Z] 10.43s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_046_barracuda
[2021-03-24T20:12:46.854Z] 10.41s call     x-pack/filebeat/tests/system/test_xpack_modules.py::XPackTest::test_fileset_file_045_barracuda
[2021-03-24T20:12:46.854Z] ======================= 323 passed in 1678.51s (0:27:58) =======================
[2021-03-24T20:12:46.854Z] >> python test: Integration Testing Complete
[2021-03-24T20:12:51.627Z] Cleaning up /var/lib/jenkins/workspace/PR-24742-1-f6718b4a-fbbd-4e25-b15c-58d565ecf674
[2021-03-24T20:12:51.628Z] Client: Docker Engine - Community
[2021-03-24T20:12:51.628Z]  Version:           20.10.3
[2021-03-24T20:12:51.628Z]  API version:       1.41
[2021-03-24T20:12:51.628Z]  Go version:        go1.13.15
[2021-03-24T20:12:51.628Z]  Git commit:        48d30b5
[2021-03-24T20:12:51.628Z]  Built:             Fri Jan 29 14:33:13 2021
[2021-03-24T20:12:51.628Z]  OS/Arch:           linux/amd64
[2021-03-24T20:12:51.628Z]  Context:           default
[2021-03-24T20:12:51.628Z]  Experimental:      true
[2021-03-24T20:12:51.628Z] 
[2021-03-24T20:12:51.628Z] Server: Docker Engine - Community
[2021-03-24T20:12:51.628Z]  Engine:
[2021-03-24T20:12:51.628Z]   Version:          20.10.3
[2021-03-24T20:12:51.628Z]   API version:      1.41 (minimum version 1.12)
[2021-03-24T20:12:51.628Z]   Go version:       go1.13.15
[2021-03-24T20:12:51.628Z]   Git commit:       46229ca
[2021-03-24T20:12:51.628Z]   Built:            Fri Jan 29 14:31:25 2021
[2021-03-24T20:12:51.628Z]   OS/Arch:          linux/amd64
[2021-03-24T20:12:51.628Z]   Experimental:     false
[2021-03-24T20:12:51.628Z]  containerd:
[2021-03-24T20:12:51.628Z]   Version:          1.4.4
[2021-03-24T20:12:51.628Z]   GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
[2021-03-24T20:12:51.628Z]  runc:
[2021-03-24T20:12:51.628Z]   Version:          1.0.0-rc93
[2021-03-24T20:12:51.628Z]   GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
[2021-03-24T20:12:51.628Z]  docker-init:
[2021-03-24T20:12:51.628Z]   Version:          0.19.0
[2021-03-24T20:12:51.628Z]   GitCommit:        de40ad0
[2021-03-24T20:12:51.628Z] Change ownership of all files inside the specific folder from root/root to current user/group
[2021-03-24T20:12:51.628Z] Unable to find image 'alpine:3.4' locally
[2021-03-24T20:12:52.198Z] 3.4: Pulling from library/alpine
[2021-03-24T20:12:52.458Z] c1e54eec4b57: Pulling fs layer
[2021-03-24T20:12:53.029Z] c1e54eec4b57: Verifying Checksum
[2021-03-24T20:12:53.029Z] c1e54eec4b57: Download complete
[2021-03-24T20:12:53.029Z] c1e54eec4b57: Pull complete
[2021-03-24T20:12:53.029Z] Digest: sha256:b733d4a32c4da6a00a84df2ca32791bb03df95400243648d8c539e7b4cce329c
[2021-03-24T20:12:53.029Z] Status: Downloaded newer image for alpine:3.4
[2021-03-24T20:12:55.574Z] Change permissions with write access of all files inside the specific folder
[2021-03-24T20:12:56.559Z] Running in /var/lib/jenkins/workspace/PR-24742-1-f6718b4a-fbbd-4e25-b15c-58d565ecf674/src/github.com/elastic/beats/build
[2021-03-24T20:12:56.922Z] + rm -rf ve
[2021-03-24T20:12:56.923Z] + find . -type d -name vendor -exec rm -r {} ;
[2021-03-24T20:12:57.384Z] + python .ci/scripts/pre_archive_test.py
[2021-03-24T20:12:59.926Z] Copy ./x-pack/filebeat/build into build/x-pack/filebeat/build
[2021-03-24T20:12:59.992Z] Running in /var/lib/jenkins/workspace/PR-24742-1-f6718b4a-fbbd-4e25-b15c-58d565ecf674/src/github.com/elastic/beats/build
[2021-03-24T20:13:00.069Z] Recording test results
[2021-03-24T20:13:04.082Z] [Checks API] No suitable checks publisher found.
[2021-03-24T20:13:04.742Z] + go clean -modcache
[2021-03-24T20:13:08.399Z] Cleaning up /var/lib/jenkins/workspace/PR-24742-1-f6718b4a-fbbd-4e25-b15c-58d565ecf674
[2021-03-24T20:13:08.399Z] Client: Docker Engine - Community
[2021-03-24T20:13:08.399Z]  Version:           20.10.3
[2021-03-24T20:13:08.399Z]  API version:       1.41
[2021-03-24T20:13:08.399Z]  Go version:        go1.13.15
[2021-03-24T20:13:08.399Z]  Git commit:        48d30b5
[2021-03-24T20:13:08.399Z]  Built:             Fri Jan 29 14:33:13 2021
[2021-03-24T20:13:08.399Z]  OS/Arch:           linux/amd64
[2021-03-24T20:13:08.399Z]  Context:           default
[2021-03-24T20:13:08.399Z]  Experimental:      true
[2021-03-24T20:13:08.399Z] 
[2021-03-24T20:13:08.399Z] Server: Docker Engine - Community
[2021-03-24T20:13:08.399Z]  Engine:
[2021-03-24T20:13:08.399Z]   Version:          20.10.3
[2021-03-24T20:13:08.399Z]   API version:      1.41 (minimum version 1.12)
[2021-03-24T20:13:08.399Z]   Go version:       go1.13.15
[2021-03-24T20:13:08.399Z]   Git commit:       46229ca
[2021-03-24T20:13:08.399Z]   Built:            Fri Jan 29 14:31:25 2021
[2021-03-24T20:13:08.399Z]   OS/Arch:          linux/amd64
[2021-03-24T20:13:08.399Z]   Experimental:     false
[2021-03-24T20:13:08.400Z]  containerd:
[2021-03-24T20:13:08.400Z]   Version:          1.4.4
[2021-03-24T20:13:08.400Z]   GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
[2021-03-24T20:13:08.400Z]  runc:
[2021-03-24T20:13:08.400Z]   Version:          1.0.0-rc93
[2021-03-24T20:13:08.400Z]   GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
[2021-03-24T20:13:08.400Z]  docker-init:
[2021-03-24T20:13:08.400Z]   Version:          0.19.0
[2021-03-24T20:13:08.400Z]   GitCommit:        de40ad0
[2021-03-24T20:13:08.400Z] Change ownership of all files inside the specific folder from root/root to current user/group
[2021-03-24T20:13:14.979Z] Change permissions with write access of all files inside the specific folder
[2021-03-24T20:13:15.615Z] Running in /var/lib/jenkins/workspace/PR-24742-1-f6718b4a-fbbd-4e25-b15c-58d565ecf674
[2021-03-24T20:13:26.773Z] Stage "Packaging" skipped due to earlier failure(s)
[2021-03-24T20:13:26.949Z] Running in /var/lib/jenkins/workspace/Beats_beats_PR-24742/src/github.com/elastic/beats
[2021-03-24T20:13:33.986Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_beats_PR-24742
[2021-03-24T20:13:34.103Z] [INFO] getVaultSecret: Getting secrets
[2021-03-24T20:13:34.219Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-03-24T20:13:35.300Z] + chmod 755 generate-build-data.sh
[2021-03-24T20:13:35.300Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-24742/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-24742/runs/1 FAILURE 6534843
[2021-03-24T20:13:35.851Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-24742/runs/1/steps/?limit=10000 -o steps-info.json
[2021-03-24T20:14:02.352Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/beats/PR-24742/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



## :bug: Flaky test report
:grey_exclamation: There are test failures but not known flaky tests.


<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

### Test stats :test_tube:
| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 3  |
| Passed       | 44774  |
| Skipped      | 4737 |
| Total        | 49514   |


### Genuine test errors [![3](https://img.shields.io/badge/3%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-24742/detail/PR-24742/1//tests)
:broken_heart: There are test failures but not known flaky tests, most likely a genuine test failure.

* **Name**: `Build&Test / filebeat-windows-2016-windows-2016 / TestDefaultHarvesterGroup/assert_a_harvester_can_be_stopped_and_removed_from_bookkeeper – github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile`
* **Name**: `Build&Test / filebeat-windows-2016-windows-2016 / TestDefaultHarvesterGroup – github.com/elastic/beats/v7/filebeat/input/filestream/internal/input-logfile`
* **Name**: `Build&Test / auditbeat-windows-8-windows-8 / TestNonRecursive – github.com/elastic/beats/v7/auditbeat/module/file_integrity/monitor`


</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

hukaixuan commented:
Face the same issue, thanks for your great job! When will this refactor be done?

jsoriano commented:
@hukaixuan thanks for your interest, I expect to dedicate some time to this in the following days, but I cannot give a fixed timeline, stay tunned to this PR :slightly_smiling_face:

jsoriano commented:
> Hey @jsoriano !
> encountered the same issue as well, appreciate your work here!
> do you have a docker image of your changes I can use to test on my end?

Hi, this is still a draft, it still doesn't cover all cases, I actually know it doesn't work well with some basic cases.

jsoriano commented on libbeat/common/kubernetes/watcher.go:
```diff
@@ -97,16 +97,15 @@ type watcher struct {
 	store = informer.GetStore()
-	queue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), objType)
-	ctx, cancel := context.WithCancel(context.Background())
+	queue = workqueue.New()
```
Not sure why a rate limited queue is needed here. Removing this for simplification, but happy to move the change to a different PR if preferred.

elasticmachine commented:
Pinging @elastic/integrations (Team:Integrations)

jsoriano commented:
Opening for a first round of reviews, but I still see some weird things with `data.ports` and with pods with multiple containers. I have to compare with previous versions.

@ChrsMark it'd be great if you could take a look, thanks!

ChrsMark reviewed (commented):
Thank you for improving all this codebase! I left some comments but it should be ok after addressing them.

One general thought is that I see that we have a lot of logic explained in comments here and there within the code. How delays work, how and why we emit stop/start events. These all look like a wider algorithm that we have to implement this feature, and we have special cases like `PodTerminated then do that` which can only be figured out is one look into the code and try to follow the flow.

I wonder if we could gather all this logic/algorithm into a more abstract document so as to have it documented centrally there. In this we will be able to run back to it when we have any kind of questions/possible bugs etc instead of interpreting the code ad-hoc. This might add an extra cost of maintaining this document but its up to us to decide to add this extra cost in order to have something useful for the future both for us but for upcoming people.
This is just a proposal since we are on it right now and can be added in another PR so as to avoid blocking this one.

exekias commented:
@vjsamuel FYI

jsoriano commented on libbeat/autodiscover/providers/kubernetes/pod.go:
```diff
@@ -284,130 +262,123 @@ func getContainersInPod(pod *kubernetes.Pod) ([]kubernetes.Container, []kubernet
 				"id":         eventID,
 				flag:         true,
-				"host":       host,
-				"port":       0,
```
This port is still set to zero some lines below if [the container is running](https://github.com/elastic/beats/pull/24742/files#diff-ce319a56cef2bfe39d4095a22160b27ec8e769020ad2de71c31ae9f3e85e912bR357), and [there are no ports defined](https://github.com/elastic/beats/pull/24742/files#diff-ce319a56cef2bfe39d4095a22160b27ec8e769020ad2de71c31ae9f3e85e912bR373).

Good catch in any case, I overlooked heartbeat autodiscover, I'll check that it continues working.

ChrsMark replied:
👍🏼

jsoriano replied:
Tested with heartbeat and it seems to be working. I found an issue, but also affects master, opened a different PR for this: https://github.com/elastic/beats/pull/25073

jsoriano commented:
Going back to draft, I am reorganizing the code so the long `emit` function is splitted in easier to document functions.

jsoriano commented on libbeat/autodiscover/providers/kubernetes/pod.go:
```diff
@@ -114,59 +114,35 @@ func NewPodEventer(uuid uuid.UUID, cfg *common.Config, client k8s.Interface, pub
+	// Stop configs in case a re-sync generates an add event for a completed pod.
+	// Completed pods can happen with cronjobs for example, they are finished but not deleted
+	// by default.
+	// For this case this will always trigger a delayed stop, that will be executed after the
```
I have reverted changes here, they are not relevant for what this PR attempts to solve.

jsoriano commented on libbeat/autodiscover/providers/kubernetes/pod.go:
```diff
@@ -284,130 +262,123 @@ func getContainersInPod(pod *kubernetes.Pod) ([]kubernetes.Container, []kubernet
+		cp := containerInPod{spec: c}
+		containersInPod[i] = &cp
+
+		// This nested loop can be a problem in pods with hundreds of containers, does it exist?
```
Refactored to use a map again to avoid the worst case scenario, and for loops to avoid generating so many temporary objects, as proposed in https://github.com/elastic/beats/pull/24742#discussion_r611388219.

Sorting the arrays is actually a good suggestion but complicates things a little bit, and can be also too much for the common use case.

We can revisit this function in a follow up if we see that needs to be improved.

ChrsMark replied:
👍🏼

jsoriano commented on libbeat/common/kubernetes/watcher.go:
```diff
@@ -216,53 +216,44 @@ func (w *watcher) enqueue(obj interface{}, state string) {
+	var entry *item
+	var ok bool
+	if entry, ok = obj.(*item); !ok {
+		// w.queue.Forget(obj)
```
More or less :slightly_smiling_face: This would be needed with the rate limiting queue, I left it so I don't forget if we go back to use it. Removing it.

jsoriano commented:
Opening again for review. Changes since previous reviews:
* `emit` has been further refactored so the main logic is hopefully more clear. Events generation has been moved out to specific functions to improve readability. Comments have been added with the intention to explain what each one of these functions do (not how they do it).
* `getContainersInPod` refactored to avoid worst-case scenario of problems with pods with many containers, and to avoid the need of a temporary array with all the containers.
* Change in `OnAdd` has been reverted, we can consider handling this corner case later if needed. Not directly related to the issue this PR intends to fix. At the moment this will only have the risk of leaving inactive harvesters running till the pods are definitely deleted.

> I wonder if we could gather all this logic/algorithm into a more abstract document so as to have it documented centrally there. In this we will be able to run back to it when we have any kind of questions/possible bugs etc instead of interpreting the code ad-hoc.  This might add an extra cost of maintaining this document but its up to us to decide to add this extra cost in order to have something useful for the future both for us and for the upcoming people.
> This is just a proposal since we are on it right now and can be added in another PR so as to avoid blocking this one.

@ChrsMark let me know if current version of this logic is easier to understand. I would like to avoid having to document this because then we have two problems: complex logic, and the need to keep documentation updated to this logic :slightly_smiling_face:

jsoriano commented:
/test

ChrsMark commented on libbeat/autodiscover/providers/kubernetes/pod.go:
```diff
@@ -257,194 +224,276 @@ func (p *pod) Stop() {
+		// so templates that need network don't generate a config.
+		if c.status.State.Running != nil {
+			if port.Name != "" && port.ContainerPort != 0 {
+				event["ports"] = common.MapStr{port.Name: port.ContainerPort}
```
🤔 Will this just keep only the last "accessed" port of the loop because we create a new  `MapStr` each time?

jsoriano replied:
Yes, I have changed the behaviour here, I have moved `data.ports` from the pod event to each container/port event. Let me know if you think this can be a problem.

Before, `data.ports` was included only in the pod event, so when configurations matched they didn't include the container metadata, only the pod metadata.

Code in this PR doesn't set the whole collection of `data.ports` in the pod event, it sets a single `data.ports.<portname>` field in to each container/port event, so when configurations match they include also the container metadata. By doing this each event with a named port is going to contain a single port.

This also fixes something I found misleading during my tests. As `data.ports` was included only in pod events, templates with conditions matching container fields, like the following one, didn't work. They work now with the changes here.
```
      templates:
        - condition:
            contains.kubernetes.container.image: redis
          config:
            - module: redis
              metricsets: ["info", "keyspace"]
              hosts: "${data.host}:${data.ports.redis}"
```

jsoriano replied:
You can see more clearly how this is different now in these test cases: https://github.com/elastic/beats/pull/24742/files#diff-9f24ba274864b933a726842c2b9fb8ee487a13545579938f9e373d0463611142R557

jsoriano replied:
Ok, what doesn't work now is this case. I will revert this change and do some more tests.
```
      templates:
        - condition:
            contains.kubernetes.pod.name: redis
          config:
            - module: redis
              hosts: "${data.host}:${data.ports.redis}"
            - module: nginx
              hosts: "${data.host}:${data.ports.http}"
```

We need more testing to cover all these cases :persevere:

ChrsMark replied:
Ah I see, it makes sense!

jsoriano replied:
Reverted in 330182725ac779372175f3ec897451935bbe6081, I will follow with this in a separate PR.

ChrsMark reviewed (commented):
Overall looks good so far 🍻 :
1. spotted one tricky line that might be a bug (maybe not :)). I might need to have a second pass with a clearer mind for hidden bug hunting (I believe there are none though)
2. @jroriano do you think you can improve testing notes a bit? I guess you have some specific scenarios you used to test this one. Would be nice to have them logged in this PR both for manual testing but also for future reference maybe.

jsoriano commented:
>   2\. @jroriano do you think you can improve testing notes a bit? I guess you have some specific scenarios you used to test this one. Would be nice to have them logged in this PR both for manual testing but also for future reference maybe.

@ChrsMark I have some quite simple kubernetes manifests, that represent some cases in a list in a private manual test case. I would like to translate this at some moment to system or [e2e](https://github.com/elastic/e2e-testing) tests.

jsoriano commented:
/test

ChrsMark reviewed (approved):
lgtm, thank you!

ChrsMark commented:
/test

exekias commented on libbeat/autodiscover/providers/kubernetes/pod.go:
```diff
@@ -257,194 +224,285 @@ func (p *pod) Stop() {
-	statuses = append(statuses, pod.Status.EphemeralContainerStatuses...)
 
-	return containers, statuses
+	delay := (flag == "stop" && podTerminated(pod, containers))
+	p.publishAll(eventList, delay)
```
does it make sense to publish a `stop` event for not terminated pods?

exekias replied:
nevermind, this would be an update, got it

exekias reviewed (approved):
LGTM! let's make sure we test this thoroughly before releasing 😇

jsoriano commented:
@ChrsMark @exekias Thanks for the reviews!

I did some additional tests and I found that the issue was not actually solved for hints-based autodiscover in filebeat, c9a3515 was missing. Could any of you please a quick review of this latest change?

jsoriano commented:
/test
